### PR TITLE
Check project against valid and deprecated trove classifiers.

### PR DESCRIPTION
**Pull request by kdeldycke**
Inspects trove classifiers on `check` CLI command calls, and look for unrecognized and deprecated categories.

Adds dependency https://github.com/pypa/trove-classifiers, a package published and maintained by PyPA that is cataloguing all classifiers. This is the canonical source of all trove definitions.

# Pull Request Check List

Resolves: #2579

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

### CODE REVIEW

**Comment by kdeldycke** [comment]
This PR is failing on Python 2.7 because it adds a dependency on [`calver`](https://github.com/di/calver), which requires `python >=  3.5`.

I guess this PR should be tied to Poetry 1.2 release then: https://github.com/python-poetry/poetry/pull/2683

**Comment by abn** [comment]
@kdeldycke unfortunately yes, if you really want it sooner you will need to disable the code for `PY2` and make the dependency optional. That said, it is unlikely this will make `1.1.0` anyway.

**Comment by kdeldycke** [comment]
Thanks @abn for the feedback! I'll wait for the next release then! :)

**Comment by kdeldycke** [comment]
I think this PR is ready to be merged, now that I rebased it on top of master, i.e. the future 1.2 release.

Unfortunately some tests are failing for what looks like something that has nothing to do with the code changes in this PR... 😞

**Comment by kdeldycke** [comment]
All checks have passed, this PR is ready for review! :)

**Comment by kdeldycke** [comment]
I just rebased this PR on top of upstream and fixed all remaining issues. All tests are passing.

This PR is ready to be merged for Poetry 1.2.

File: src/poetry/console/commands/check.py [hunk 1]
```diff
@@ -19,6 +25,32 @@ def handle(self) -> int:
...
+            check_result["errors"].append(
+                f"Unrecognized classifiers: {unrecognized!r}."
+            )
+
+        deprecated = sorted(
+            set(config.get("classifiers", [])).intersection(set(deprecated_classifiers))
+        )
+        if deprecated:
+            for old_classifier in deprecated:
+                message = f"Deprecated classifier {old_classifier!r}."
```
Comment (neersighted) [inline src/poetry/console/commands/check.py:44]: Does it make more sense to print the repr (`!r`) here than the string form, and if so, why?

File: src/poetry/console/commands/check.py [hunk 2]
```diff
@@ -19,6 +25,32 @@ def handle(self) -> int:
         poetry_file = Factory.locate(Path.cwd())
         config = PyProjectTOML(poetry_file).poetry_config
         check_result = Factory.validate(config, strict=True)
+
+        unrecognized = sorted(
+            set(config.get("classifiers", []))
```
Comment (neersighted) [inline src/poetry/console/commands/check.py:30]: It's small but I'd rather only run `set(<value>)` and `config.get()` once.

File: src/poetry/console/commands/check.py [hunk 3]
```diff
@@ -19,6 +25,32 @@ def handle(self) -> int:
...
+                f"Unrecognized classifiers: {unrecognized!r}."
+            )
+
+        deprecated = sorted(
+            set(config.get("classifiers", [])).intersection(set(deprecated_classifiers))
+        )
+        if deprecated:
+            for old_classifier in deprecated:
+                message = f"Deprecated classifier {old_classifier!r}."
+                new_classifiers = cast(
```
Comment (neersighted) [inline src/poetry/console/commands/check.py:45]: Really not a fan of this inline cast -- I think we can find a better way to write this to still satisfy Mypy.

File: src/poetry/console/commands/check.py [hunk 4]
```diff
@@ -19,6 +25,32 @@ def handle(self) -> int:
...
+                    Dict[str, List[str]], deprecated_classifiers[old_classifier]
+                )
+                if new_classifiers:
+                    message += f" Must be replaced by {new_classifiers!r}."
```
Comment (neersighted) [inline src/poetry/console/commands/check.py:49]: I'd rather build a complete sentence than the multiple here.

**Review by neersighted — CHANGES_REQUESTED** [review]
This feature looks good overall, but I have some code-level changes I'd like to see. Additionally, I would like to see this code factored out into a helper rather than inline in the check command, and some more thought given to variable naming to help keep it maintainable later.

File: src/poetry/console/commands/check.py (hunk 3 above)
Comment (dimbleby) [reply src/poetry/console/commands/check.py:45]: it's not correct either is it?  `deprecated_classifiers` is the dictionary, `new_classifiers` is just a list of strings?

Comment (dimbleby) [reply src/poetry/console/commands/check.py:45]: https://github.com/pypa/trove-classifiers/pull/109

Comment (dimbleby) [reply src/poetry/console/commands/check.py:45]: which has been merged and auto-released, so trove-classifiers 2022.6.23 should make the cast redundant

Comment (kdeldycke) [reply src/poetry/console/commands/check.py]: Just removed the explicit casting and bumped `trove-classifiers` to `2022.6.23` in https://github.com/python-poetry/poetry/pull/2881/commits/404998ce080a6bd2b5d5729813d13df8061d44ea

File: src/poetry/console/commands/check.py (hunk 1 above)
Comment (kdeldycke) [reply src/poetry/console/commands/check.py:41]: The fully-qualified classifier is a string delimited by ` :: ` separators. To make the error message more readable I wanted to have visual clues to materialize the start and end of the classifier string. That way the user can easely copy and paste the classifier string from the error message while reducing mistakes pertaining to extra spaces from the full error message.

I chose `!r` for consistency with the `{new_classifiers!r}` message below, which is a list of classifiers. That way I'm certain all classifiers strings will be rendered with the same kind of quotes (i.e. simple tick: `'`).

Do you want me to add that explanation as comment in the source code?

File: src/poetry/console/commands/check.py (hunk 2 above)
Comment (kdeldycke) [reply src/poetry/console/commands/check.py]: Good catch! Fixed in: https://github.com/python-poetry/poetry/pull/2881/commits/2eb0354983d49ca7a4d4aead37f3e2d9cc1e82d3

File: src/poetry/console/commands/check.py (hunk 4 above)
Comment (kdeldycke) [reply src/poetry/console/commands/check.py]: Fixed in: https://github.com/python-poetry/poetry/pull/2881/commits/d940d2440a323afe77c275037d1c7fdd53f6a0f5

File: src/poetry/console/commands/check.py (hunk 1 above)
Comment (kdeldycke) [reply src/poetry/console/commands/check.py]: I added that explanation into the classifier helper docstring at: https://github.com/python-poetry/poetry/pull/2881/commits/d1f7e5ca2c5734a29ad022475d1f3baa30a6e5be

**Comment by kdeldycke** [comment]
All concerns identified during the code review have been addressed. All tests passes. This PR is ready to be merged.

Do you want me to squash the commits into one?

> This feature looks good overall, but I have some code-level changes I'd like to see. Additionally, I would like to see this code factored out into a helper rather than inline in the check command, and some more thought given to variable naming to help keep it maintainable later.

As for these structural code changes, I regrouped all the code into a naive helper in: https://github.com/python-poetry/poetry/pull/2881/commits/89eba1e9c598118e8cdf490271e00fb716b8092e

Is that what you were looking for? Do you have suggestions on the variables to rename?

**Comment by neersighted** [comment]
Hey @kdeldycke -- sorry about the wait on this one, I never commented here, but it was concluded that this wasn't something we wanted to include in 1.2. I've moved the milestone to 1.3, and hope to have this reviewed and merged in time for the next minor release.

**Review by neersighted — APPROVED** [review]

**Comment by kdeldycke** [comment]
1.3 is a perfectly fine target! Thanks @neersighted for the merge and sorry for the late reply!

**Pull request merged**